README.md
light Cookbook - [![Cookbook Version](https://img.shields.io/cookbook/v/light.svg)](https://community.opscode.com/cookbooks/light) [![Build Status](https://travis-ci.org/sliim-cookbooks/light.svg?branch=master)](https://travis-ci.org/sliim-cookbooks/light)
================
The light Cookbook - Installs [light](https://github.com/haikarainen/light)

Requirements
------------
#### Platforms
The following platforms and versions are tested and supported using Opscode's test-kitchen.
- `Debian 9`
- `Ubuntu 16.04`

Attributes
----------
#### light::default
| Key                    | Type   | Description                                                        |
| ---------------------- | ------ | ------------------------------------------------------------------ |
| `[light][install_dir]` | String | Installation directory (default: `/opt/light`)                     |
| `[light][repository]`  | String | light repository (default: `https://github.com/haikarainen/light`) |
| `[light][reference]`   | String | light repository branch or tag (default: `v1.0`)                   |
| `[light][env]`         | Hash   | environment for git resource (default: `{}`)                       |

Usage
-----
#### light::default
Include the `light` recipe in your run_list to install light:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[light]"
  ]
}
```

Testing
-------
See [TESTING.md](TESTING.md)

Contributing
------------
See [CONTRIBUTING.md](CONTRIBUTING.md)

## License and Authors

Authors: Sliim <sliim@mailoo.org>

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.



CHANGELOG.md
light CHANGELOG
===============

This file is used to list changes made in each version of the light cookbook.

1.0.0
-----
- Initial release of light cookbook


spec/default_spec.rb
# -*- coding: utf-8 -*-

require 'chefspec'
require 'chefspec/berkshelf'

describe 'light::default' do
  let(:subject) do
    ChefSpec::SoloRunner.new(platform: 'debian', version: '9.9') do |node|
      node.override['light']['install_dir'] = '/opt/light'
      node.override['light']['repository'] = 'https://custom/repo.git'
      node.override['light']['reference'] = 'custom-ref'
      node.override['light']['env'] = { 'CUSTOM' => 'custom-env' }
    end.converge(described_recipe)
  end

  it 'installs package[help2man]' do
    expect(subject).to install_package('help2man')
  end

  it 'syncs git[/opt/light]' do
    expect(subject).to sync_git('/opt/light')
      .with(repository: 'https://custom/repo.git',
            reference: 'custom-ref',
            environment: { 'CUSTOM' => 'custom-env' })
  end

  ['/etc/light/save', '/etc/light/mincap'].each do |dir|
    it "creates directory[#{dir}]" do
      expect(subject).to create_directory(dir)
        .with(owner: 'root',
              group: 'root',
              mode: '0755',
              recursive: true)
    end
  end

  it 'creates link[/usr/local/bin/light]' do
    expect(subject).to create_link('/usr/local/bin/light')
      .with(owner: 'root',
            group: 'root',
            to: '/opt/light/light')
  end

  it 'creates directory[/etc/udev/rules.d]' do
    expect(subject).to create_directory('/etc/udev/rules.d')
      .with(owner: 'root',
            group: 'root',
            mode: '0755',
            recursive: true)
  end

  it 'creates file[/etc/udev/rules.d/90-backlight.rules]' do
    expect(subject).to create_file('/etc/udev/rules.d/90-backlight.rules')
      .with(owner: 'root',
            group: 'root',
            mode: '0640')
  end
end


test/integration/default/default_spec.rb
# -*- coding: utf-8 -*-

['/opt/light', '/etc/light/save', '/etc/light/mincap'].each do |dir|
  describe file dir do
    it { should exist }
    it { should be_directory }
    it { should be_owned_by 'root' }
  end
end

describe file '/opt/light/light' do
  it { should exist }
  it { should be_file }
  it { should be_executable }
  it { should be_owned_by 'root' }
end

describe file '/usr/local/bin/light' do
  it { should exist }
  it { should be_linked_to '/opt/light/light' }
end

describe file '/etc/udev/rules.d/90-backlight.rules' do
  it { should exist }
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
end


recipes/default.rb
# -*- coding: utf-8 -*-
#
# Cookbook::light
# Recipe::default
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

install_dir = node['light']['install_dir']

package 'help2man'

execute 'make-light' do
  command 'make'
  cwd install_dir
  action :nothing
end

git install_dir do
  repository node['light']['repository']
  reference node['light']['reference']
  environment node['light']['env'] unless node['light']['env'].empty?
  action :sync
  notifies :run, 'execute[make-light]'
end

['/etc/light/save', '/etc/light/mincap'].each do |dir|
  directory dir do
    recursive true
    owner 'root'
    group 'root'
    mode '0755'
  end
end

link '/usr/local/bin/light' do
  owner 'root'
  group 'root'
  to "#{install_dir}/light"
end

directory '/etc/udev/rules.d' do
  owner 'root'
  group 'root'
  mode '0755'
  recursive true
end

file '/etc/udev/rules.d/90-backlight.rules' do
  owner 'root'
  group 'root'
  mode '0640'
  content ['ACTION=="add", SUBSYSTEM=="backlight", RUN+="/bin/chgrp video /sys/class/backlight/%k/brightness"',
           'ACTION=="add", SUBSYSTEM=="backlight", RUN+="/bin/chmod g+w /sys/class/backlight/%k/brightness"'].join("\n")
end
